File: iosApp/iosApp/blockchain/events/commutoswap/SwapFilledEvent.swift
//
//  SwapFilledEvent.swift
//  iosApp
//
//  Created by jimmyt on 9/5/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import BigInt
import web3swift

/**
 Represents a [SwapFilled](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#swapfilled) event, emitted by the CommutoSwap smart contract when a swap is filled.
 */
class SwapFilledEvent {
    /**
     The ID of the filled swap.
     */
    let id: UUID
    /**
     The ID of the blockchain on which the event was emitted.
     */
    let chainID: BigUInt
    /**
     The hash of the transaction that emitted this event, as a lowercae hex string with a "0x" prefix.
     */
    let transactionHash: String
    /**
     Creates a new `SwapFilledEvent` given a web3swift `EventParserResultProtocol` containing information from a [SwapFilled](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#swapfilled) event, or returns `nil` if the passed `EventParserResultProtocol` doesn't contain information from a SwapFilled event.
     */
    init?(_ result: EventParserResultProtocol, chainID: BigUInt) {
        guard let resultID = UUID.from(data: result.decodedResult["swapID"] as? Data) else {
            return nil
        }
        id = resultID
        self.chainID = chainID
        guard var transactionHash = result.transactionReceipt?.transactionHash.toHexString().lowercased() else {
            return nil
        }
        if !transactionHash.starts(with: "0x") {
            transactionHash = "0x" + transactionHash
        }
        self.transactionHash = transactionHash
    }
    /**
     Creates a new `SwapFilledEvent` with the given `id` and `chainID`.
     */
    init(id: UUID, chainID: BigUInt, transactionHash: String) {
        self.id = id
        self.chainID = chainID
        self.transactionHash = transactionHash
    }
    
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/blockchain/events/commutoswap/SwapFilledEvent.kt
package com.commuto.interfacemobile.android.blockchain.events.commutoswap

import com.commuto.interfacemobile.android.contractwrapper.CommutoSwap
import java.math.BigInteger
import java.nio.ByteBuffer
import java.util.*

/**
 * Represents a [SwapFilled](https://www.commuto.xyz/docs/technical-reference/core-tec-ref#swapfilled) event, emitted by
 * the CommutoSwap smart contract when a swap is filled.
 *
 * @property swapID The ID of the filled swap.
 * @property chainID The ID of the blockchain on which this event was emitted.
 * @property transactionHash The hash of the transaction that emitted this event, as a lowercase hex string with a "0x"
 * prefix.
 */
class SwapFilledEvent(val swapID: UUID, val chainID: BigInteger, transactionHash: String) {

    val transactionHash: String

    init {
        this.transactionHash = if (transactionHash.startsWith("0x")) {
            transactionHash.lowercase()
        } else {
            "0x${transactionHash.lowercase()}"
        }
    }

    companion object {
        /**
         * Creates a [SwapFilledEvent] from a [CommutoSwap.SwapFilledEventResponse] and a specified [chainID].
         *
         * @param event The [CommutoSwap.SwapFilledEventResponse] containing the swap ID of a filled swap.
         * @param chainID The ID of the blockchain on which this event was emitted.
         *
         * @return A new [SwapFilledEvent] with a swap ID as a [UUID] derived from [event]'s swap ID [ByteArray] and the
         * specified [chainID], and the transaction hash specified by [event].
         */
        fun fromEventResponse(event: CommutoSwap.SwapFilledEventResponse, chainID: BigInteger): SwapFilledEvent {
            val swapIDByteBuffer = ByteBuffer.wrap(event.swapID)
            val mostSigBits = swapIDByteBuffer.long
            val leastSigBits = swapIDByteBuffer.long
            return SwapFilledEvent(UUID(mostSigBits, leastSigBits), chainID, event.log.transactionHash)
        }

    }

}